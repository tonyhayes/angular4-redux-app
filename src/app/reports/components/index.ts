import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ReactiveFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';

import { ReportInfoComponent } from './report-info';
import { ReportDetailComponent } from './report-detail';
import { ReportPreviewComponent } from './report-preview';
import { ReportPreviewListComponent } from './report-preview-list';
import { ReportSearchComponent } from './report-search';

import { PipesModule } from '../../shared/pipes';
import { MaterialModule } from '../../material';

export const COMPONENTS = [
  ReportInfoComponent,
  ReportDetailComponent,
  ReportPreviewComponent,
  ReportPreviewListComponent,
  ReportSearchComponent,
];

@NgModule({
  imports: [
    CommonModule,
    ReactiveFormsModule,
    MaterialModule,
    RouterModule,
    PipesModule,
  ],
  declarations: COMPONENTS,
  exports: COMPONENTS,
})
export class ComponentsModule {}
