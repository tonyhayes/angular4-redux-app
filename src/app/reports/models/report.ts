export interface ScheduleInfo {
  name: string; /*Lists your schedules by the name you assign to them*/
  repeats: RepeatTime; /*Describes how often the schedule is set to run.*/
  sourceFile: string; /*file associated with the schedule*/
  outputLocation: string; /* location that the scheduled report is saved*/
  lastRun: string; /* last time and date when the schedule was run*/
  nextRun: string; /* next time and date when the schedule will run again*/
  status: string; /*current Status of the schedule. The state can be either Normal or Paused.*/
}
export interface ReportInfo {
  name: string; /*Lists your report by the name you assign to them*/
  sourceFile: string; /*file associated with the report*/
  outputLocation: string; /* location that the report is saved*/
  lastRun: string; /* last time and date when the report was run*/
  status: string; /*current Status of the report. The state can be either Normal or Paused.*/
}

export interface RepeatTime {
  repeatTimeStart: string; /* all schedules are blocked from running from date*/
  repeatTimeEnds: string; /* all schedules are blocked from running to date*/
  repeats: string; /*Describes how often backout is active.*/
  repeatsEndBy: string; /*Describes when the  repeater ends.*/
}

export interface Report {
  id: string;
  blackoutTimesList: RepeatTime[];
  reportInfo: ReportInfo;
  scheduleInfo: ScheduleInfo;
}

export function generateMockReport(): Report {
  return {
    id: '1',
    blackoutTimesList: [
      {
        repeatTimeStart: '1513616909621' ,
        repeatTimeEnds: '1513616909621',
        repeats: 'Daily',
        repeatsEndBy: 'Never'
      }
    ],
    reportInfo: {
      name: 'report name',
      sourceFile: '/path/to/report',
      outputLocation: '/output/path/to/report',
      lastRun: '1513616909621',
      status: 'Normal',
    },
    scheduleInfo: {
      name: 'report schedule name',
      repeats: {
        repeatTimeStart: '1513616909621',
        repeatTimeEnds: '1513616909621',
        repeats: 'Daily',
        repeatsEndBy: 'Never'
      },
      sourceFile: '/path/to/report',
      outputLocation: '/output/path/to/report',
      lastRun: '1513616909621',
      nextRun: '1513616909621',
      status: 'Normal',
    },
  };
}
