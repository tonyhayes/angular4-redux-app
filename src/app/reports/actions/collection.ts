import { Action } from '@ngrx/store';
import { Report } from '../models/report';

export const ADD_REPORT = '[Collection] Add Report';
export const ADD_REPORT_SUCCESS = '[Collection] Add Report Success';
export const ADD_REPORT_FAIL = '[Collection] Add Report Fail';
export const REMOVE_REPORT = '[Collection] Remove Report';
export const REMOVE_REPORT_SUCCESS = '[Collection] Remove Report Success';
export const REMOVE_REPORT_FAIL = '[Collection] Remove Report Fail';
export const LOAD = '[Collection] Load';
export const LOAD_SUCCESS = '[Collection] Load Success';
export const LOAD_FAIL = '[Collection] Load Fail';

/**
 * Add Report to Collection Actions
 */
export class AddReport implements Action {
  readonly type = ADD_REPORT;

  constructor(public payload: Report) {}
}

export class AddReportSuccess implements Action {
  readonly type = ADD_REPORT_SUCCESS;

  constructor(public payload: Report) {}
}

export class AddReportFail implements Action {
  readonly type = ADD_REPORT_FAIL;

  constructor(public payload: Report) {}
}

/**
 * Remove Report from Collection Actions
 */
export class RemoveReport implements Action {
  readonly type = REMOVE_REPORT;

  constructor(public payload: Report) {}
}

export class RemoveReportSuccess implements Action {
  readonly type = REMOVE_REPORT_SUCCESS;

  constructor(public payload: Report) {}
}

export class RemoveReportFail implements Action {
  readonly type = REMOVE_REPORT_FAIL;

  constructor(public payload: Report) {}
}

/**
 * Load Collection Actions
 */
export class Load implements Action {
  readonly type = LOAD;
}

export class LoadSuccess implements Action {
  readonly type = LOAD_SUCCESS;

  constructor(public payload: Report[]) {}
}

export class LoadFail implements Action {
  readonly type = LOAD_FAIL;

  constructor(public payload: any) {}
}

export type Actions =
  | AddReport
  | AddReportSuccess
  | AddReportFail
  | RemoveReport
  | RemoveReportSuccess
  | RemoveReportFail
  | Load
  | LoadSuccess
  | LoadFail;
