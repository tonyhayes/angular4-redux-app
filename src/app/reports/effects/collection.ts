import 'rxjs/add/operator/map';
import 'rxjs/add/operator/catch';
import 'rxjs/add/operator/switchMap';
import 'rxjs/add/operator/mergeMap';
import 'rxjs/add/operator/toArray';
import { Injectable } from '@angular/core';
import { Action } from '@ngrx/store';
import { Effect, Actions } from '@ngrx/effects';
import { Database } from '@ngrx/db';
import { Observable } from 'rxjs/Observable';
import { defer } from 'rxjs/observable/defer';
import { of } from 'rxjs/observable/of';

import * as collection from '../actions/collection';
import { Report } from '../models/report';

@Injectable()
export class CollectionEffects {
  /**
   * This effect does not yield any actions back to the store. Set
   * `dispatch` to false to hint to @ngrx/effects that it should
   * ignore any elements of this effect stream.
   *
   * The `defer` observable accepts an observable factory function
   * that is called when the observable is subscribed to.
   * Wrapping the database open call in `defer` makes
   * effect easier to test.
   */
  @Effect({ dispatch: false })
  openDB$: Observable<any> = defer(() => {
    return this.db.open('reports_app');
  });

  @Effect()
  loadCollection$: Observable<Action> = this.actions$
    .ofType(collection.LOAD)
    .switchMap(() =>
      this.db
        .query('reports')
        .toArray()
        .map((reports: Report[]) => new collection.LoadSuccess(reports))
        .catch(error => of(new collection.LoadFail(error)))
    );

  @Effect()
  addReportToCollection$: Observable<Action> = this.actions$
    .ofType(collection.ADD_REPORT)
    .map((action: collection.AddReport) => action.payload)
    .mergeMap(report =>
      this.db
        .insert('reports', [report])
        .map(() => new collection.AddReportSuccess(report))
        .catch(() => of(new collection.AddReportFail(report)))
    );

  @Effect()
  removeReportFromCollection$: Observable<Action> = this.actions$
    .ofType(collection.REMOVE_REPORT)
    .map((action: collection.RemoveReport) => action.payload)
    .mergeMap(report =>
      this.db
        .executeWrite('reports', 'delete', [report.id])
        .map(() => new collection.RemoveReportSuccess(report))
        .catch(() => of(new collection.RemoveReportFail(report)))
    );

  constructor(private actions$: Actions, private db: Database) {}
}
