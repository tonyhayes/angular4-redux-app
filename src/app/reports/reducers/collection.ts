import * as collection from '../actions/collection';

export interface State {
  loaded: boolean;
  loading: boolean;
  ids: string[];
}

const initialState: State = {
  loaded: false,
  loading: false,
  ids: [],
};

export function reducer(
  state = initialState,
  action: collection.Actions
): State {
  switch (action.type) {
    case collection.LOAD: {
      return {
        ...state,
        loading: true,
      };
    }

    case collection.LOAD_SUCCESS: {
      return {
        loaded: true,
        loading: false,
        ids: action.payload.map(report => report.id),
      };
    }

    case collection.ADD_REPORT_SUCCESS:
    case collection.REMOVE_REPORT_FAIL: {
      if (state.ids.indexOf(action.payload.id) > -1) {
        return state;
      }

      return {
        ...state,
        ids: [...state.ids, action.payload.id],
      };
    }

    case collection.REMOVE_REPORT_SUCCESS:
    case collection.ADD_REPORT_FAIL: {
      return {
        ...state,
        ids: state.ids.filter(id => id !== action.payload.id),
      };
    }

    default: {
      return state;
    }
  }
}

export const getLoaded = (state: State) => state.loaded;

export const getLoading = (state: State) => state.loading;

export const getIds = (state: State) => state.ids;
