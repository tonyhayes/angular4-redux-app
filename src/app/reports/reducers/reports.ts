import { createSelector } from '@ngrx/store';
import { createEntityAdapter, EntityAdapter, EntityState } from '@ngrx/entity';
import { Report } from '../models/report';
import * as report from '../actions/report';
import * as collection from '../actions/collection';

/**
 * @ngrx/entity provides a predefined interface for handling
 * a structured dictionary of records. This interface
 * includes an array of ids, and a dictionary of the provided
 * model type by id. This interface is extended to include
 * any additional interface properties.
 */
export interface State extends EntityState<Report> {
  selectedReportId: string | null;
}

/**
 * createEntityAdapter creates many an object of helper
 * functions for single or multiple operations
 * against the dictionary of records. The configuration
 * object takes a record id selector function and
 * a sortComparer option which is set to a compare
 * function if the records are to be sorted.
 */
export const adapter: EntityAdapter<Report> = createEntityAdapter<Report>({
  selectId: (report: Report) => report.id,
  sortComparer: false,
});

/**
 * getInitialState returns the default initial state
 * for the generated entity state. Initial state
 * additional properties can also be defined.
 */
export const initialState: State = adapter.getInitialState({
  selectedReportId: null,
});

export function reducer(
  state = initialState,
  action: report.Actions | collection.Actions
): State {
  switch (action.type) {
    case report.SEARCH_COMPLETE:
    case collection.LOAD_SUCCESS: {
      return {
        /**
         * The addMany function provided by the created adapter
         * adds many records to the entity dictionary
         * and returns a new state including those records. If
         * the collection is to be sorted, the adapter will
         * sort each record upon entry into the sorted array.
         */
        ...adapter.addMany(action.payload, state),
        selectedReportId: state.selectedReportId,
      };
    }

    case report.LOAD: {
      return {
        /**
         * The addOne function provided by the created adapter
         * adds one record to the entity dictionary
         * and returns a new state including that records if it doesn't
         * exist already. If the collection is to be sorted, the adapter will
         * insert the new record into the sorted array.
         */
        ...adapter.addOne(action.payload, state),
        selectedReportId: state.selectedReportId,
      };
    }

    case report.SELECT: {
      return {
        ...state,
        selectedReportId: action.payload,
      };
    }

    default: {
      return state;
    }
  }
}

/**
 * Because the data structure is defined within the reducer it is optimal to
 * locate our selector functions at this level. If store is to be thought of
 * as a database, and reducers the tables, selectors can be considered the
 * queries into said database. Remember to keep your selectors small and
 * focused so they can be combined and composed to fit each particular
 * use-case.
 */

export const getSelectedId = (state: State) => state.selectedReportId;
