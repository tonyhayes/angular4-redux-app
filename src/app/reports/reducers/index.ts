import { createSelector, createFeatureSelector } from '@ngrx/store';
import * as fromSearch from './search';
import * as fromReports from './reports';
import * as fromCollection from './collection';
import * as fromRoot from '../../reducers';

export interface ReportsState {
  search: fromSearch.State;
  reports: fromReports.State;
  collection: fromCollection.State;
}

export interface State extends fromRoot.State {
  'reports': ReportsState;
}

export const reducers = {
  search: fromSearch.reducer,
  reports: fromReports.reducer,
  collection: fromCollection.reducer,
};

/**
 * A selector function is a map function factory. We pass it parameters and it
 * returns a function that maps from the larger state tree into a smaller
 * piece of state. This selector simply selects the `reports` state.
 *
 * Selectors are used with the `select` operator.
 *
 * ```ts
 * class MyComponent {
 * 	constructor(state$: Observable<State>) {
 * 	  this.reportsState$ = state$.select(getReportsState);
 * 	}
 * }
 * ```
 */

/**
 * The createFeatureSelector function selects a piece of state from the root of the state object.
 * This is used for selecting feature states that are loaded eagerly or lazily.
*/
export const getReportsState = createFeatureSelector<ReportsState>('reports');

/**
 * Every reducer module exports selector functions, however child reducers
 * have no knowledge of the overall state tree. To make them usable, we
 * need to make new selectors that wrap them.
 *
 * The createSelector function creates very efficient selectors that are memoized and
 * only recompute when arguments change. The created selectors can also be composed
 * together to select different pieces of state.
 */
export const getReportEntitiesState = createSelector(
  getReportsState,
  state => state.reports
);

export const getSelectedReportId = createSelector(
  getReportEntitiesState,
  fromReports.getSelectedId
);

/**
 * Adapters created with @ngrx/entity generate
 * commonly used selector functions including
 * getting all ids in the record set, a dictionary
 * of the records by id, an array of records and
 * the total number of records. This reducers boilerplate
 * in selecting records from the entity state.
 */
export const {
  selectIds: getReportIds,
  selectEntities: getReportEntities,
  selectAll: getAllReports,
  selectTotal: getTotalReports,
} = fromReports.adapter.getSelectors(getReportEntitiesState);

export const getSelectedReport = createSelector(
  getReportEntities,
  getSelectedReportId,
  (entities, selectedId) => {
    return selectedId && entities[selectedId];
  }
);

/**
 * Just like with the reports selectors, we also have to compose the search
 * reducer's and collection reducer's selectors.
 */
export const getSearchState = createSelector(
  getReportsState,
  (state: ReportsState) => state.search
);

export const getSearchReportIds = createSelector(
  getSearchState,
  fromSearch.getIds
);
export const getSearchQuery = createSelector(
  getSearchState,
  fromSearch.getQuery
);
export const getSearchLoading = createSelector(
  getSearchState,
  fromSearch.getLoading
);
export const getSearchError = createSelector(
  getSearchState,
  fromSearch.getError
);

/**
 * Some selector functions create joins across parts of state. This selector
 * composes the search result IDs to return an array of reports in the store.
 */
export const getSearchResults = createSelector(
  getReportEntities,
  getSearchReportIds,
  (reports, searchIds) => {
    return searchIds.map(id => reports[id]);
  }
);

export const getCollectionState = createSelector(
  getReportsState,
  (state: ReportsState) => state.collection
);

export const getCollectionLoaded = createSelector(
  getCollectionState,
  fromCollection.getLoaded
);
export const getCollectionLoading = createSelector(
  getCollectionState,
  fromCollection.getLoading
);
export const getCollectionReportIds = createSelector(
  getCollectionState,
  fromCollection.getIds
);

export const getReportCollection = createSelector(
  getReportEntities,
  getCollectionReportIds,
  (entities, ids) => {
    return ids.map(id => entities[id]);
  }
);

export const isSelectedReportInCollection = createSelector(
  getCollectionReportIds,
  getSelectedReportId,
  (ids, selected) => {
    return ids.indexOf(selected) > -1;
  }
);
