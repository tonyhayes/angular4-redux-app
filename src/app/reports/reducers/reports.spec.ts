import { reducer } from './reports';
import * as fromReports from './reports';
import { SearchComplete, Load, Select } from '../actions/report';
import { Report, generateMockReport } from '../models/report';
import { LoadSuccess } from '../actions/collection';

describe('ReportsReducer', () => {
  const report1 = generateMockReport();
  const report2 = { ...report1, id: '222' };
  const report3 = { ...report1, id: '333' };
  const initialState: fromReports.State = {
    ids: [report1.id, report2.id],
    entities: {
      [report1.id]: report1,
      [report2.id]: report2,
    },
    selectedReportId: null,
  };

  describe('undefined action', () => {
    it('should return the default state', () => {
      const result = reducer(undefined, {} as any);

      expect(result).toMatchSnapshot();
    });
  });

  describe('SEARCH_COMPLETE & LOAD_SUCCESS', () => {
    function noExistingReports(
      action: any,
      reportsInitialState: any,
      initialState: any,
      reports: Report[]
    ) {
      const createAction = new action(reports);

      const result = reducer(reportsInitialState, createAction);

      expect(result).toMatchSnapshot();
    }

    function existingReports(action: any, initialState: any, reports: Report[]) {
      // should not replace existing reports
      const differentReport2 = { ...reports[0], foo: 'bar' };
      const createAction = new action([reports[1], differentReport2]);

      const expectedResult = {
        ids: [...initialState.ids, reports[1].id],
        entities: {
          ...initialState.entities,
          [reports[1].id]: reports[1],
        },
        selectedReportId: null,
      };

      const result = reducer(initialState, createAction);

      expect(result).toMatchSnapshot();
    }

    it('should add all reports in the payload when none exist', () => {
      noExistingReports(SearchComplete, fromReports.initialState, initialState, [
        report1,
        report2,
      ]);

      noExistingReports(LoadSuccess, fromReports.initialState, initialState, [
        report1,
        report2,
      ]);
    });

    it('should add only new reports when reports already exist', () => {
      existingReports(SearchComplete, initialState, [report2, report3]);

      existingReports(LoadSuccess, initialState, [report2, report3]);
    });
  });

  describe('LOAD', () => {
    const expectedResult = {
      ids: [report1.id],
      entities: {
        [report1.id]: report1,
      },
      selectedReportId: null,
    };

    it('should add a single report, if the report does not exist', () => {
      const action = new Load(report1);

      const result = reducer(fromReports.initialState, action);

      expect(result).toMatchSnapshot();
    });

    it('should return the existing state if the report exists', () => {
      const action = new Load(report1);

      const result = reducer(expectedResult, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('SELECT', () => {
    it('should set the selected report id on the state', () => {
      const action = new Select(report1.id);

      const result = reducer(initialState, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('Selectors', () => {
    describe('getSelectedId', () => {
      it('should return the selected id', () => {
        const result = fromReports.getSelectedId({
          ...initialState,
          selectedReportId: report1.id,
        });

        expect(result).toMatchSnapshot();
      });
    });
  });
});
