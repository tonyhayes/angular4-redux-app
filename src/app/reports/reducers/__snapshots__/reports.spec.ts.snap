// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportsReducer LOAD should add a single report, if the report does not exist 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "blackoutTimesList": Array [
        Object {
          "repeatTimeEnds": "1513616909621",
          "repeatTimeStart": "1513616909621",
          "repeats": "Daily",
          "repeatsEndBy": "Never",
        },
      ],
      "id": "1",
      "reportInfo": Object {
        "lastRun": "1513616909621",
        "name": "report name",
        "outputLocation": "/output/path/to/report",
        "sourceFile": "/path/to/report",
        "status": "Normal",
      },
      "scheduleInfo": Object {
        "lastRun": "1513616909621",
        "name": "report schedule name",
        "nextRun": "1513616909621",
        "outputLocation": "/output/path/to/report",
        "repeats": Object {
          "repeatTimeEnds": "1513616909621",
          "repeatTimeStart": "1513616909621",
          "repeats": "Daily",
          "repeatsEndBy": "Never",
        },
        "sourceFile": "/path/to/report",
        "status": "Normal",
      },
    },
  },
  "ids": Array [
    "1",
  ],
  "selectedReportId": null,
}
`;

exports[`ReportsReducer LOAD should return the existing state if the report exists 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "blackoutTimesList": Array [
        Object {
          "repeatTimeEnds": "1513616909621",
          "repeatTimeStart": "1513616909621",
          "repeats": "Daily",
          "repeatsEndBy": "Never",
        },
      ],
      "id": "1",
      "reportInfo": Object {
        "lastRun": "1513616909621",
        "name": "report name",
        "outputLocation": "/output/path/to/report",
        "sourceFile": "/path/to/report",
        "status": "Normal",
      },
      "scheduleInfo": Object {
        "lastRun": "1513616909621",
        "name": "report schedule name",
        "nextRun": "1513616909621",
        "outputLocation": "/output/path/to/report",
        "repeats": Object {
          "repeatTimeEnds": "1513616909621",
          "repeatTimeStart": "1513616909621",
          "repeats": "Daily",
          "repeatsEndBy": "Never",
        },
        "sourceFile": "/path/to/report",
        "status": "Normal",
      },
    },
  },
  "ids": Array [
    "1",
  ],
  "selectedReportId": null,
}
`;

exports[`ReportsReducer SEARCH_COMPLETE & LOAD_SUCCESS should add all reports in the payload when none exist 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "blackoutTimesList": Array [
        Object {
          "repeatTimeEnds": "1513616909621",
          "repeatTimeStart": "1513616909621",
          "repeats": "Daily",
          "repeatsEndBy": "Never",
        },
      ],
      "id": "1",
      "reportInfo": Object {
        "lastRun": "1513616909621",
        "name": "report name",
        "outputLocation": "/output/path/to/report",
        "sourceFile": "/path/to/report",
        "status": "Normal",
      },
      "scheduleInfo": Object {
        "lastRun": "1513616909621",
        "name": "report schedule name",
        "nextRun": "1513616909621",
        "outputLocation": "/output/path/to/report",
        "repeats": Object {
          "repeatTimeEnds": "1513616909621",
          "repeatTimeStart": "1513616909621",
          "repeats": "Daily",
          "repeatsEndBy": "Never",
        },
        "sourceFile": "/path/to/report",
        "status": "Normal",
      },
    },
    "222": Object {
      "blackoutTimesList": Array [
        Object {
          "repeatTimeEnds": "1513616909621",
          "repeatTimeStart": "1513616909621",
          "repeats": "Daily",
          "repeatsEndBy": "Never",
        },
      ],
      "id": "222",
      "reportInfo": Object {
        "lastRun": "1513616909621",
        "name": "report name",
        "outputLocation": "/output/path/to/report",
        "sourceFile": "/path/to/report",
        "status": "Normal",
      },
      "scheduleInfo": Object {
        "lastRun": "1513616909621",
        "name": "report schedule name",
        "nextRun": "1513616909621",
        "outputLocation": "/output/path/to/report",
        "repeats": Object {
          "repeatTimeEnds": "1513616909621",
          "repeatTimeStart": "1513616909621",
          "repeats": "Daily",
          "repeatsEndBy": "Never",
        },
        "sourceFile": "/path/to/report",
        "status": "Normal",
      },
    },
  },
  "ids": Array [
    "1",
    "222",
  ],
  "selectedReportId": null,
}
`;

exports[`ReportsReducer SEARCH_COMPLETE & LOAD_SUCCESS should add all reports in the payload when none exist 2`] = `
Object {
  "entities": Object {
    "1": Object {
      "blackoutTimesList": Array [
        Object {
          "repeatTimeEnds": "1513616909621",
          "repeatTimeStart": "1513616909621",
          "repeats": "Daily",
          "repeatsEndBy": "Never",
        },
      ],
      "id": "1",
      "reportInfo": Object {
        "lastRun": "1513616909621",
        "name": "report name",
        "outputLocation": "/output/path/to/report",
        "sourceFile": "/path/to/report",
        "status": "Normal",
      },
      "scheduleInfo": Object {
        "lastRun": "1513616909621",
        "name": "report schedule name",
        "nextRun": "1513616909621",
        "outputLocation": "/output/path/to/report",
        "repeats": Object {
          "repeatTimeEnds": "1513616909621",
          "repeatTimeStart": "1513616909621",
          "repeats": "Daily",
          "repeatsEndBy": "Never",
        },
        "sourceFile": "/path/to/report",
        "status": "Normal",
      },
    },
    "222": Object {
      "blackoutTimesList": Array [
        Object {
          "repeatTimeEnds": "1513616909621",
          "repeatTimeStart": "1513616909621",
          "repeats": "Daily",
          "repeatsEndBy": "Never",
        },
      ],
      "id": "222",
      "reportInfo": Object {
        "lastRun": "1513616909621",
        "name": "report name",
        "outputLocation": "/output/path/to/report",
        "sourceFile": "/path/to/report",
        "status": "Normal",
      },
      "scheduleInfo": Object {
        "lastRun": "1513616909621",
        "name": "report schedule name",
        "nextRun": "1513616909621",
        "outputLocation": "/output/path/to/report",
        "repeats": Object {
          "repeatTimeEnds": "1513616909621",
          "repeatTimeStart": "1513616909621",
          "repeats": "Daily",
          "repeatsEndBy": "Never",
        },
        "sourceFile": "/path/to/report",
        "status": "Normal",
      },
    },
  },
  "ids": Array [
    "1",
    "222",
  ],
  "selectedReportId": null,
}
`;

exports[`ReportsReducer SEARCH_COMPLETE & LOAD_SUCCESS should add only new reports when reports already exist 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "blackoutTimesList": Array [
        Object {
          "repeatTimeEnds": "1513616909621",
          "repeatTimeStart": "1513616909621",
          "repeats": "Daily",
          "repeatsEndBy": "Never",
        },
      ],
      "id": "1",
      "reportInfo": Object {
        "lastRun": "1513616909621",
        "name": "report name",
        "outputLocation": "/output/path/to/report",
        "sourceFile": "/path/to/report",
        "status": "Normal",
      },
      "scheduleInfo": Object {
        "lastRun": "1513616909621",
        "name": "report schedule name",
        "nextRun": "1513616909621",
        "outputLocation": "/output/path/to/report",
        "repeats": Object {
          "repeatTimeEnds": "1513616909621",
          "repeatTimeStart": "1513616909621",
          "repeats": "Daily",
          "repeatsEndBy": "Never",
        },
        "sourceFile": "/path/to/report",
        "status": "Normal",
      },
    },
    "222": Object {
      "blackoutTimesList": Array [
        Object {
          "repeatTimeEnds": "1513616909621",
          "repeatTimeStart": "1513616909621",
          "repeats": "Daily",
          "repeatsEndBy": "Never",
        },
      ],
      "id": "222",
      "reportInfo": Object {
        "lastRun": "1513616909621",
        "name": "report name",
        "outputLocation": "/output/path/to/report",
        "sourceFile": "/path/to/report",
        "status": "Normal",
      },
      "scheduleInfo": Object {
        "lastRun": "1513616909621",
        "name": "report schedule name",
        "nextRun": "1513616909621",
        "outputLocation": "/output/path/to/report",
        "repeats": Object {
          "repeatTimeEnds": "1513616909621",
          "repeatTimeStart": "1513616909621",
          "repeats": "Daily",
          "repeatsEndBy": "Never",
        },
        "sourceFile": "/path/to/report",
        "status": "Normal",
      },
    },
    "333": Object {
      "blackoutTimesList": Array [
        Object {
          "repeatTimeEnds": "1513616909621",
          "repeatTimeStart": "1513616909621",
          "repeats": "Daily",
          "repeatsEndBy": "Never",
        },
      ],
      "id": "333",
      "reportInfo": Object {
        "lastRun": "1513616909621",
        "name": "report name",
        "outputLocation": "/output/path/to/report",
        "sourceFile": "/path/to/report",
        "status": "Normal",
      },
      "scheduleInfo": Object {
        "lastRun": "1513616909621",
        "name": "report schedule name",
        "nextRun": "1513616909621",
        "outputLocation": "/output/path/to/report",
        "repeats": Object {
          "repeatTimeEnds": "1513616909621",
          "repeatTimeStart": "1513616909621",
          "repeats": "Daily",
          "repeatsEndBy": "Never",
        },
        "sourceFile": "/path/to/report",
        "status": "Normal",
      },
    },
  },
  "ids": Array [
    "1",
    "222",
    "333",
  ],
  "selectedReportId": null,
}
`;

exports[`ReportsReducer SEARCH_COMPLETE & LOAD_SUCCESS should add only new reports when reports already exist 2`] = `
Object {
  "entities": Object {
    "1": Object {
      "blackoutTimesList": Array [
        Object {
          "repeatTimeEnds": "1513616909621",
          "repeatTimeStart": "1513616909621",
          "repeats": "Daily",
          "repeatsEndBy": "Never",
        },
      ],
      "id": "1",
      "reportInfo": Object {
        "lastRun": "1513616909621",
        "name": "report name",
        "outputLocation": "/output/path/to/report",
        "sourceFile": "/path/to/report",
        "status": "Normal",
      },
      "scheduleInfo": Object {
        "lastRun": "1513616909621",
        "name": "report schedule name",
        "nextRun": "1513616909621",
        "outputLocation": "/output/path/to/report",
        "repeats": Object {
          "repeatTimeEnds": "1513616909621",
          "repeatTimeStart": "1513616909621",
          "repeats": "Daily",
          "repeatsEndBy": "Never",
        },
        "sourceFile": "/path/to/report",
        "status": "Normal",
      },
    },
    "222": Object {
      "blackoutTimesList": Array [
        Object {
          "repeatTimeEnds": "1513616909621",
          "repeatTimeStart": "1513616909621",
          "repeats": "Daily",
          "repeatsEndBy": "Never",
        },
      ],
      "id": "222",
      "reportInfo": Object {
        "lastRun": "1513616909621",
        "name": "report name",
        "outputLocation": "/output/path/to/report",
        "sourceFile": "/path/to/report",
        "status": "Normal",
      },
      "scheduleInfo": Object {
        "lastRun": "1513616909621",
        "name": "report schedule name",
        "nextRun": "1513616909621",
        "outputLocation": "/output/path/to/report",
        "repeats": Object {
          "repeatTimeEnds": "1513616909621",
          "repeatTimeStart": "1513616909621",
          "repeats": "Daily",
          "repeatsEndBy": "Never",
        },
        "sourceFile": "/path/to/report",
        "status": "Normal",
      },
    },
    "333": Object {
      "blackoutTimesList": Array [
        Object {
          "repeatTimeEnds": "1513616909621",
          "repeatTimeStart": "1513616909621",
          "repeats": "Daily",
          "repeatsEndBy": "Never",
        },
      ],
      "id": "333",
      "reportInfo": Object {
        "lastRun": "1513616909621",
        "name": "report name",
        "outputLocation": "/output/path/to/report",
        "sourceFile": "/path/to/report",
        "status": "Normal",
      },
      "scheduleInfo": Object {
        "lastRun": "1513616909621",
        "name": "report schedule name",
        "nextRun": "1513616909621",
        "outputLocation": "/output/path/to/report",
        "repeats": Object {
          "repeatTimeEnds": "1513616909621",
          "repeatTimeStart": "1513616909621",
          "repeats": "Daily",
          "repeatsEndBy": "Never",
        },
        "sourceFile": "/path/to/report",
        "status": "Normal",
      },
    },
  },
  "ids": Array [
    "1",
    "222",
    "333",
  ],
  "selectedReportId": null,
}
`;

exports[`ReportsReducer SELECT should set the selected report id on the state 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "blackoutTimesList": Array [
        Object {
          "repeatTimeEnds": "1513616909621",
          "repeatTimeStart": "1513616909621",
          "repeats": "Daily",
          "repeatsEndBy": "Never",
        },
      ],
      "id": "1",
      "reportInfo": Object {
        "lastRun": "1513616909621",
        "name": "report name",
        "outputLocation": "/output/path/to/report",
        "sourceFile": "/path/to/report",
        "status": "Normal",
      },
      "scheduleInfo": Object {
        "lastRun": "1513616909621",
        "name": "report schedule name",
        "nextRun": "1513616909621",
        "outputLocation": "/output/path/to/report",
        "repeats": Object {
          "repeatTimeEnds": "1513616909621",
          "repeatTimeStart": "1513616909621",
          "repeats": "Daily",
          "repeatsEndBy": "Never",
        },
        "sourceFile": "/path/to/report",
        "status": "Normal",
      },
    },
    "222": Object {
      "blackoutTimesList": Array [
        Object {
          "repeatTimeEnds": "1513616909621",
          "repeatTimeStart": "1513616909621",
          "repeats": "Daily",
          "repeatsEndBy": "Never",
        },
      ],
      "id": "222",
      "reportInfo": Object {
        "lastRun": "1513616909621",
        "name": "report name",
        "outputLocation": "/output/path/to/report",
        "sourceFile": "/path/to/report",
        "status": "Normal",
      },
      "scheduleInfo": Object {
        "lastRun": "1513616909621",
        "name": "report schedule name",
        "nextRun": "1513616909621",
        "outputLocation": "/output/path/to/report",
        "repeats": Object {
          "repeatTimeEnds": "1513616909621",
          "repeatTimeStart": "1513616909621",
          "repeats": "Daily",
          "repeatsEndBy": "Never",
        },
        "sourceFile": "/path/to/report",
        "status": "Normal",
      },
    },
  },
  "ids": Array [
    "1",
    "222",
  ],
  "selectedReportId": "1",
}
`;

exports[`ReportsReducer Selectors getSelectedId should return the selected id 1`] = `"1"`;

exports[`ReportsReducer undefined action should return the default state 1`] = `
Object {
  "entities": Object {},
  "ids": Array [],
  "selectedReportId": null,
}
`;
